src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -625,13 +658,20 @@ pub struct CollectionState<T> {
     /// The last reported timestamp bindings, if any.
     /// This is used to differentialize timestamp bindings received before storing them in stash
     pub(super) last_reported_ts_bindings: HashMap<PartitionId, MzOffset>,
+
+    /// A `ReadHandle` for the backing persist shard/collection. This internally holds back the
+    /// since frontier and we need to downgrade that when the read capabilities change.
+    // TODO(aljoscha): Once all sources are wired up to go through persist/STORAGE, this will stop
+    // being optional
+    pub read_handle: Option<Box<dyn CollectionReadHandle<T>>>,
```

Comment:
The `Controller` type that implement `StorageController` already directly talks to `persist`. Is there a reason to hide this behind a trait?

Reply:
Adding a `ReadHandle` to this structs "infects" all the types that contain it with `T: Timestamp + Lattice + Codec64` constraints. It's probably a problem in the first place that `ReadHandle` has those on the struct itself, but I thought it would be nice anyways to encapsulate.